crates/memory/src/op_store/time_slice_hash_store.rs

```diff
@@ -0,0 +1,532 @@
+        assert_eq!(5, store.inner.len());
+
+        assert_eq!(
+            SliceHash::Single {
+                id: 100,
+                hash: vec![1, 2, 3].into()
+            },
+            store.inner[1]
+        );
+        assert_eq!(
+            SliceHash::Single {
+                id: 101,
+                hash: vec![3, 4, 5].into()
+            },
+            store.inner[2]
+        );
+        assert_eq!(
+            SliceHash::Single {
+                id: 102,
+                hash: vec![2, 3, 4].into()
+            },
+            store.inner[3]
+        );
+
+        assert!(store.check());
+    }
+
+    #[test]
+    fn overwrite_existing_single_hash() {
+        let mut store = TimeSliceHashStore::new();
+
+        store.insert(100, vec![1, 2, 3].into()).unwrap();
+        assert_eq!(3, store.inner.len());
+
+        assert_eq!(
+            SliceHash::Single {
+                id: 100,
+                hash: vec![1, 2, 3].into()
+            },
+            store.inner[1]
+        );
+
+        store.insert(100, vec![2, 3, 4].into()).unwrap();
+        assert_eq!(3, store.inner.len());
+
+        assert_eq!(
+            SliceHash::Single {
+                id: 100,
+                hash: vec![2, 3, 4].into()
+            },
+            store.inner[1]
+        );
+
+        assert!(store.check());
+    }
+
+    #[test]
+    fn highest_stored_ignores_empty() {
+        let mut store = TimeSliceHashStore::new();
+
+        store.insert(100, vec![1, 2, 3].into()).unwrap();
+        assert_eq!(Some(100), store.highest_stored_id);
+
+        store.insert(5000, bytes::Bytes::new()).unwrap();
+        assert_eq!(Some(100), store.highest_stored_id);
+
+        store.insert(120, vec![2, 3, 4].into()).unwrap();
+        assert_eq!(Some(120), store.highest_stored_id);
+
+        assert!(store.check());
+    }
+
+    #[test]
+    fn get() {
+        let mut store = TimeSliceHashStore::new();
+
+        store.insert(100, vec![1, 2, 3].into()).unwrap();
+
+        assert_eq!(
+            Some(&SliceHash::Single {
+                id: 100,
+                hash: vec![1, 2, 3].into()
+            }),
+            store.get(100)
+        );
+        assert_eq!(Some(&SliceHash::Block { id: 0, count: 99 }), store.get(0));
+        assert_eq!(Some(&SliceHash::Block { id: 0, count: 99 }), store.get(30));
+        assert_eq!(Some(&SliceHash::Block { id: 0, count: 99 }), store.get(99));
+        assert_eq!(
+            Some(&SliceHash::Block {
+                id: 101,
+                count: u64::MAX - 101
+            }),
+            store.get(101)
+        );
+        assert_eq!(
+            Some(&SliceHash::Block {
+                id: 101,
+                count: u64::MAX - 101
+            }),
+            store.get(500)
+        );
+        assert_eq!(
+            Some(&SliceHash::Block {
+                id: 101,
+                count: u64::MAX - 101
+            }),
+            store.get(u64::MAX)
+        );
+    }
+}
```

Comment:
do we want tests about how empty hashes get handled?

Reply:
Yes, now tested by `insert_empty_hash_into_empty`